function/avro-import.py

```diff
@@ -0,0 +1,156 @@
+    except urllib.error.URLError as ue:
+        return handle_exception(job_id, start_time, "URL Error - the url " + url + " is not valid.")
+
+    try:
+        reader = fastavro.reader(avro)
+        translation = translate(reader, options)
+        metadata = {
+            "namespace": workspace_namespace,
+            "name": workspace_name,
+            "userSubjectId": user_subject_id,
+            "userEmail": user_email,
+            "jobId": job_id,
+            "startTime": start_time,
+        }
+        metadata_json_str = str(json.dumps(metadata))
+        metadat_file_name = job_id + "/metadata.json"
+        upsert_json_str = str(json.dumps(translation))
+        upsert_file_name = job_id + "/upsert.json"
+        write_to_bucket(metadat_file_name, metadata_json_str)
+        write_to_bucket(upsert_file_name, upsert_json_str)
+    except Exception as e:
+        return handle_exception(job_id, start_time, "The following exception occurred: " + str(e))
+    except:
+        return handle_exception(job_id, start_time, "Something went wrong with the following request: " + str(request.json))
+
+
+def handle_exception(subdirectory_name, start_time, message):
+    error_file_name = subdirectory_name + "/error-" + start_time + ".txt"
+    write_to_bucket(error_file_name, message)
+    return message
+
+
+def write_upsert_to_bucket(job_id, content_string):
+    compressed_value = gzip.compress(bytes(content_string, 'utf-8'))
+    file_name = job_id + "/upsert.json"
+    write_to_bucket(file_name, compressed_value)
+
+
+def write_metadata_to_bucket(job_id, content_string):
+    file_name = job_id + "/metadata.json"
+    write_to_bucket(file_name, content_string)
+
+
+def write_to_bucket(file_name, content):
+    storage_client = storage.Client()
+    bucket = storage_client.get_bucket("avro-translated-json")
+    blob = bucket.blob(file_name)
+    blob.upload_from_string(content)
+
+
+def translate(reader, options=None):
+    if options is None:
+        options = {}
+    t = Translator(options)
+    return t.translate(reader)
+
+
+class Translator:
+    def __init__(self, options=None):
+        if options is None:
+            options = {}
+        defaults = {'b64-decode-enums': False, 'prefix-object-ids': False}
+        self.options = {**defaults, **options}
+
+    def translate(self, reader):
+        if reader is None:
+            return None
+
+        enums = _list_enums(reader.writer_schema)
+        results = [self._translate_record(record, enums)
+                   for record in reader if record['name'] != 'Metadata']
+        return results
+
+    def _translate_record(self, record, enums):
+        entity_type = record['name']
+        name = record['id']
+
+        def make_op(key, value):
+            if self.options['b64-decode-enums'] and (entity_type, key) in enums:
+                value = _b64_decode(value).decode("utf-8")
+            if self.options['prefix-object-ids'] and key == 'object_id':
```

Comment:
do you mean `elif` here, or do you explicitly want to overwrite the value of `value` if both are true?

Reply:
I guess it could be. If switching to `elif`, I think it would read better swapping the conditions. As written, it's "if these independent options are enabled... and then some other stuff". Swapping it would be more like "if the key matches some mutually-exclusive conditions... and an option is enabled". I think I wrote it the way I did subconsciously thinking, if the option is disabled, it will short-circuit the conditional. FWIW, there was a time when the `(entity_type, key) in enums` check was measurably more expensive because `enums` was a list instead of a dict so I had some optimization on my mind.